seismicpro/survey/survey.py

```diff
@@ -1479,3 +1525,46 @@ def construct_fold_map(self, by, id_cols=None, agg=None, bin_size=None):
         coords = tmp_map.index_data[tmp_map.coords_cols]
         values = tmp_map.index_data[tmp_map.metric_name]
         return tmp_map.metric.construct_map(coords, values, index=index, agg=agg, bin_size=bin_size)
+
+    def construct_qc_maps(self, by, metrics=None, agg=None, bin_size=None):
+        """Construct a map of tracewise metric aggregated by gathers.
+
+        Parameters
+        ----------
+        by : {"source", "shot", "receiver", "rec", "cdp", "cmp", "midpoint", "bin", "supergather"}
+            Gather type to aggregate header values over.
+        metric : str or list of str, optional
+            name(s) of metrics to build metrics maps. If None, maps for all metrics that were calculated for this
+            survey are built.
+        agg : str or callable, optional, defaults to "mean"
+            An aggregation function. Passed directly to `pandas.core.groupby.DataFrameGroupBy.agg`.
+        bin_size : int, float or array-like with length 2, optional
+            Bin size for X and Y axes. If single `int` or `float`, the same bin size will be used for both axes.
+
+        Returns
+        -------
+        BaseMetricMap
+            Constructed metric map.
+        """
+
+        squeeze_output = isinstance(metrics, str)
+        if metrics is None:
+            metrics = list(self.qc_metrics.keys())
+        metrics = to_list(metrics)
+
+        if not set(metrics) <= set(self.qc_metrics.keys()):
+            wrong_names = ', '.join(set(metrics) - set(self.qc_metrics.keys()))
+            raise ValueError(f"Metrics with name(s) ({wrong_names}) are(is) not calculated yet!")
+
+        index_cols, coords_cols = get_cols_from_by(self, by)
+        index_cols = to_list(coords_cols) if index_cols is None else to_list(index_cols)
+        coords_cols = to_list(coords_cols)
+
+        mmaps = []
+        for metric_name in metrics:
+            metric = self.qc_metrics[metric_name].provide_context(survey=self)
+            columns = set(index_cols + coords_cols + to_list(metric.header_cols))
+            metric_mmap = metric.construct_map(self.get_headers(columns), index_cols, coords_cols, agg=agg,
+                                               bin_size=bin_size)
```

Comment:
```suggestion
            columns = set(index_cols + coords_cols + to_list(metric.header_cols))
            metric_mmap = metric.construct_map(coords, self.get_headers(metric.header_cols), index=index, agg=agg,
                                               bin_size=bin_size)
```
And calculate `coords` and `index` just after `coords_cols` and `index_cols` outside the loop